Expose plain Dex configuration in CSC ConfigMap

gdemonet opened the pull request:
**Component**: salt, dex, service-config

**Context**: See #2781 

**Summary**:

- change CSC format, described by new apiVersion `addons.metalk8s.scality.com/v1alpha2`
- adjust how "defaults" are loaded (now rely on `slsutil.renderer` to allow Jinja templating, choice of rendering engine being left to the source file) so we can store and update default clients and issuer URL
- adjust Grafana role mapping: all Dex static users are Admin

_Left to do_:
- [x] Document new format
- [x] Adjust existing procedures accordingly
- [x] Add procedure to configure an OIDC connector (useful for Keycloak) - kind of done, simply referred to Dex docs
- [x] Remove pre-downgrade logic, because 2.6 is not backward compatible

**Acceptance criteria**: 

- Upgrade still works, and Dex behaviour is unchanged, even with non-default configuration in Dex CSC

    - tested with a custom password and a GitHub connector, from 2.5.1 to 2.6.0

- Grafana role mapping still works by default, and additional Dex users are also Admin
- Tests pass (of course)
- Documented procedures for configuring Dex still work

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2781 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
charts: Drop Dex Secret from rendered chart

Since we want the Secret to be entirely generated from our
ServiceConfig data and associated defaults, it doesn't make sense to
keep it in the rendered chart. We don't have an available toggle in the
chart options, so we rely on the following invocation of the render
script:

./charts/render.py dex --namespace metalk8s-auth charts/dex.yaml \
  --service-config dex metalk8s-dex-config \
                   metalk8s/addons/dex/config/dex.yaml metalk8s-auth \
  --remove-manifest Secret dex \
  charts/dex/ > salt/metalk8s/addons/dex/deployed/chart.sls

See: #2781

gdemonet committed:
salt/dex: Update ServiceConfig format and defaults

We define a new format (under the apiVersion /v1alpha2) to store the
flat Dex configuration under its `spec.config` key (akin to what is done
for Alertmanager).
We also update the initialization of this ServiceConfig to include every
item requiring some templating (defaults cannot include Jinja syntax).

See: #2781

gdemonet committed:
salt/dex: Handle conversion of ServiceConfig

When applying metalk8s.addons.dex.deployed.service-configuration, if the
existing ConfigMap is in the old format, we will automatically convert
it to the new /v1alpha2 format.
When downgrading, we will automatically convert to the old format.

Note that when performing this conversion, we ensure the previous
configuration is backed up in the same ConfigMap, under the
'previous-config.yaml' key, just in case.

See: #2781

gdemonet committed:
charts: Update Grafana list of admin users

We update the mapping of roles from OIDC users in Grafana configuration,
to match with the updated format of Dex ServiceConfig.
Note that while the previous approach was checking if an authenticated
user's email was containing (not exact match) the first static user's
email, we rework this to make sure that all configured static users in
Dex config will be granted the Admin role.

We re-render the prometheus-operator chart with:

```
./charts/render.py prometheus-operator \
  charts/prometheus-operator.yaml charts/prometheus-operator/ \
  --namespace metalk8s-monitoring \
  --service-config grafana metalk8s-grafana-config \
                   metalk8s/addons/prometheus-operator/config/grafana.yaml \
                   metalk8s-monitoring \
  --service-config prometheus metalk8s-prometheus-config \
                   metalk8s/addons/prometheus-operator/config/prometheus.yaml \
                   metalk8s-monitoring \
  --service-config alertmanager metalk8s-alertmanager-config \
                   metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
                   metalk8s-monitoring \
  --service-config dex metalk8s-dex-config \
                   metalk8s/addons/dex/config/dex.yaml \
                   metalk8s-auth \
  --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
> salt/metalk8s/addons/prometheus-operator/deployed/chart.sls
```

See: #2781

gdemonet committed:
salt/dex: Store managed config in defaults

We previously saved most of the configuration data in the user-editable
ConfigMap. However, some of this configuration needs to be managed by us
(namely, anything related to the control-plane Ingress controller), and
letting this in the ConfigMap prevents any change on our side (which is
likely to happen on upgrades or Bootstrap restore).

The defaults, previously defined in a static YAML file, are now
requiring Jinja rendering, so we slightly adjust the loading behaviour
(using `salt.slsutil.renderer` and relying on its parsing of the
"shebang" in the source file).

See: #2781

gdemonet committed:
salt, charts: Use slsutil.renderer everywhere

Instead of the previous import_yaml Jinja directive, we now rely on the
slsutil.renderer execution method, which is more generic and will let us
change defaults for CSC without changing the rendered charts or any
formula consuming it.
Note that we update the render script to use this behaviour, and
re-render all affected charts accordingly.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

TeddyAndrieux commented on salt/metalk8s/addons/logging/loki/deployed/chart.sls:
```diff
@@ -184,7 +184,7 @@ spec:
     metadata:
       annotations:
-        checksum/config: f6917e98336282f93b8cf80b99ca1e78f7adfd297b98c367c1cc420a1cb8d1ac
+        checksum/config: f9c7883a31a1ef8ef3264cfb5160f3cc015a91cf46c6687d7f2777a62af6936b
```
Not related to this PR but since we have CSC for loki configuration this one should be computed by salt.
And it's the same for alertmanager, prometheus and grafana I think (+ maybe others)

gdemonet replied:
Good point, I'll add it in this PR if that's OK for you

gdemonet replied:
So, Prometheus doesn't have a "config file" per-se, it's entirely managed through CRs by prometheus-operator, so I don't have something to compute a checksum from.
Did it for Grafana and Alertmanager though (even if the latter is also managed by prometheus-operator), and Loki of course.

gdemonet committed:
charts: Add config checksum for Grafana and AM

Since both services are using a CSC-derived configuration, we add a
Salt `__slot__` to make sure their Pods are restarted whenever we
re-render the controllers after a configuration update.

Charts were updated using:

./charts/render.py prometheus-operator \
  charts/prometheus-operator.yaml charts/prometheus-operator/ \
  --namespace metalk8s-monitoring \
  --service-config grafana metalk8s-grafana-config \
                   metalk8s/addons/prometheus-operator/config/grafana.yaml \
                   metalk8s-monitoring \
  --service-config prometheus metalk8s-prometheus-config \
                   metalk8s/addons/prometheus-operator/config/prometheus.yaml \
                   metalk8s-monitoring \
  --service-config alertmanager metalk8s-alertmanager-config \
                   metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
                   metalk8s-monitoring \
  --service-config dex metalk8s-dex-config \
                   metalk8s/addons/dex/config/dex.yaml.j2 \
                   metalk8s-auth \
  --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
> salt/metalk8s/addons/prometheus-operator/deployed/chart.sls

gdemonet committed:
charts: Add config checksum for Loki

Loki configuration is derived from a CSC ConfigMap, thus we add a Salt
`__slot__` to make sure its Pods will be updated properly when this
configuration changes.

Chart was re-rendered using the following:

./charts/render.py loki charts/loki.yaml charts/loki/ \
  --remove-manifest Secret loki \
  --namespace metalk8s-logging \
  --service-config loki metalk8s-loki-config \
                   metalk8s/addons/logging/loki/config/loki.yaml \
                   metalk8s-logging \
> salt/metalk8s/addons/logging/loki/deployed/chart.sls

gdemonet committed:
doc: Update CSC defaults inclusion

We now have a Saltstack "renderer shebang" at the beginning of these
files, which may not make sense for a user.
We thus hide these, and also hide some comments in the recently updated
Dex defaults.

See: #2781

gdemonet committed:
doc: Update Dex-related configuration procedures

Since we recently updated the CSC format for Dex, we need to update the
corresponding procedures.
Note that we also move the role binding guide out of the CSC-specific
procedures, and into the "Account Administration" page, which should be
easier for users to find.

See: #2781

gdemonet committed:
doc: Add section to disable the static user store

This user store is not perfect, and doesn't benefit from the same
features as a full-fledged identity management system would. As such,
it may make sense for users of MetalK8s to disable it, and only use Dex
as an intermediary between their identity provider(s) and MetalK8s
services.

See: #2781

gdemonet committed:
doc: Add section for other Dex configurations

Since we now support the full Dex config file format, we simply link to
Dex documentation and explain the generic procedure for editing the CSC
and propagating changes.

See: #2781

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2781-plain-dex-csc` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2781-plain-dex-csc`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2781-plain-dex-csc
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2781-plain-dex-csc
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort










*The following options are set:* **approve**

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -306,27 +316,62 @@ For more details on Alert Rules, see the official
+To disable (resp. enable) it, perform the following steps:
 
-   .. _Generate-password-hash:
+#. Set the ``enablePasswordDB`` configuration flag to ``true`` (resp.
```
```suggestion
#. Set the ``enablePasswordDB`` configuration flag to ``false`` (resp.
```

gdemonet replied:
:facepalm:

gdemonet replied:
Done.

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -335,180 +380,169 @@ To add a new static user, perform the following operations:
 
       root@bootstrap $ kubectl --kubeconfig /etc/kubernetes/admin.conf \
-                         apply -f ClusterRoleBinding.yaml
+                         edit configmaps metalk8s-dex-config -n metalk8s-auth
```
To be consistent with the change you made above:
```suggestion
                         edit configmap metalk8s-dex-config -n metalk8s-auth
```

gdemonet replied:
Will do

gdemonet replied:
Done.

alexandre-allard commented on docs/operation/cluster_and_service_configuration.rst:
```diff
@@ -335,180 +380,169 @@ To add a new static user, perform the following operations:
    .. code-block:: shell
 
       root@bootstrap $ kubectl --kubeconfig /etc/kubernetes/admin.conf \
                          edit configmaps metalk8s-dex-config -n metalk8s-auth
```
ditto

gdemonet replied:
Done.

alexandre-allard commented:
Not tested, only found details in the documentation part, otherwise LGTM.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye gdemonet.
